### DISCUSSION TITLE
JK-B2A8S20P UART connection: Missing "balancing switch" and "Balance current Ah" on Version 11 ( HW: 11.XW SW 11.287)

### DISCUSSION THREAD

**Discussion by ldiegos** [syssi/esphome-jk-bms#462] [Q&A] [answered]
Good night,

I'm trying to simulate the jk bt application on home assistant and i realize that i miss the balancing switch, to turn off and on the balancer, it seems to be part of it, as the pic shows... I wish to have the same switch as the Charging and Discharging ones:
![image](https://github.com/syssi/esphome-jk-bms/assets/29803600/cee89551-1bbf-43bd-b6e8-30400cbb7fe0)

Regarding the Balance Current Ah I miss it also in the module, nothing seems to be the value of the balancer amphs.
![image](https://github.com/syssi/esphome-jk-bms/assets/29803600/fb242a15-ff2a-4da3-a43e-b6c3b34a78f7)

Are they ready implemented? I follow the discussion https://github.com/syssi/esphome-jk-bms/discussions/374 and i would like to know if those are already implemented for version 11 ( HW: 11.XW SW 11.287)

In the other hand everything works fine in the JK-B2A8S20P versión 11 ( HW: 11.XW SW 11.287) with UART and esp32.

Thanks a lot!!!
Luis.

**Comment by syssi** [comment, marked as answer]
The UART protocol doesn't contain the balancer current measurement. It can be retrieved via Bluetooth only. It's unknown and may be not possible to control the balancer switch using UART too.

**Reply by ldiegos** [reply]
What a pitty... thanks so much!!!

**Reply by Jackzerox** [reply]
Hi, Im trying to use the yaml uart and ble configuration, but at the end, the compiling send me this menssage

Linking .pioenvs/espjkbms/firmware.elf
RAM:   [==        ]  16.4% (used 53780 bytes from 327680 bytes)
Flash: [==========]  101.7% (used 1866277 bytes from 1835008 bytes)
Error: The program size (1866277 bytes) is greater than maximum allowed (1835008 bytes)
*** [checkprogsize] Explicit exit, status 1

**Reply by syssi** [reply]
@Jackzerox please create a new issue and provide your YAML configuration and specify your used ESP model.

**Comment by thitcher** [comment]
Is is it the same behaviour with cell resistance?
Only over BLE?

**Reply by syssi** [reply]
Yes. Cell resistance is only available via BLE.

**Reply by thitcher** [reply]
Oh, OK.
So I have to try if the BLE Signal comes through my metal-box where the Batteries placed in.
Is it possible to read JK-BMS and NEEY Balancer with one ESP over BLE?

**Reply by syssi** [reply]
Yes. This is possible. If you have the chance you should avoid a WiFi connection and pick a dev board equipped with an (by ESPHome supported) ethernet module.

**Reply by thitcher** [reply]
THX for the reply.
Do you have an example file that shows me how to code 1 JK-BMS and 1 NEEY together?
I´ll found examples with more than one bms, but nothing in combination with neey.

**Reply by thitcher** [reply]
I tried to merge the code to use it for 1x JK-BMS and 1x NEEY Balancer.
But I get these errorcodes:

`INFO ESPHome 2024.2.2
INFO Reading configuration /config/esphome/esphome-web-c8145c.yaml...
INFO Updating https://github.com/syssi/esphome-jk-bms.git@main
Failed config

jk_bms_ble: [source /config/esphome/esphome-web-c8145c.yaml:94]
  - ble_client_id: client0
    protocol_version: JK02_32S
    throttle: 5s
    id: bms
    update_interval: 5s
  
  'protocol_version' is a required option for [1].
  - ble_client_id: client1
    throttle: 5s
    id: bal
binary_sensor.heltec_balancer_ble: [source /config/esphome/esphome-web-c8145c.yaml:114]
  
  Component binary_sensor.heltec_balancer_ble requires component heltec_balancer_ble.
  platform: heltec_balancer_ble
  heltec_balancer_ble: bal
  balancing: 
    name: neey bal balancing
  online_status: 
    name: neey bal online status
button.heltec_balancer_ble: [source /config/esphome/esphome-web-c8145c.yaml:129]
  
  Component button.heltec_balancer_ble requires component heltec_balancer_ble.
  platform: heltec_balancer_ble
  heltec_balancer_ble_id: bal
  retrieve_settings: 
    name: neey bal retrieve settings
    id: retrieve_settings_button_neey
  retrieve_device_info: 
    name: neey bal retrieve device info
  retrieve_factory_defaults: 
    name: neey bal retrieve factory defaults
number.heltec_balancer_ble: [source /config/esphome/esphome-web-c8145c.yaml:150]
  
  Component number.heltec_balancer_ble requires component heltec_balancer_ble.
  platform: heltec_balancer_ble
  heltec_balancer_ble_id: bal
  cell_count: 
    name: neey bal cell count
  balance_trigger_voltage: 
    name: neey bal balance trigger voltage
  max_balance_current: 
    name: neey bal max balance current
  balance_sleep_voltage: 
    name: neey bal balance sleep voltage
  balance_start_voltage: 
    name: neey bal balance start voltage
  nominal_battery_capacity: 
    name: neey bal nominal battery capacity
switch.heltec_balancer_ble: [source /config/esphome/esphome-web-c8145c.yaml:326]
  
  Component switch.heltec_balancer_ble requires component heltec_balancer_ble.
  platform: heltec_balancer_ble
  heltec_balancer_ble_id: bal
  balancer: 
    name: neey bal balancer
text_sensor.heltec_balancer_ble: [source /config/esphome/esphome-web-c8145c.yaml:348]
  
  Component text_sensor.heltec_balancer_ble requires component heltec_balancer_ble.
  platform: heltec_balancer_ble
  heltec_balancer_ble_id: bal
  operation_status: 
    name: neey bal operation status
  total_runtime_formatted: 
    name: neey bal total runtime formatted
  buzzer_mode: 
    name: neey bal buzzer mode
  battery_type: 
    name: neey bal battery type`

Can you please help me to get it to work?
At the moment I try it with wifi, if it compiles correct, I will switch to LAN
Here is my code:
[merge.yaml.txt](https://github.com/syssi/esphome-jk-bms/files/14569617/merge.yaml.txt)
